File: frontend/src/lib/nlsql.ts
import { api } from "./api";

export interface SchemaColumn {
  name: string;
  type: string;
}

export interface SchemaTable {
  table: string;
  columns: SchemaColumn[];
}

export interface NLQueryResponse {
  sql: string | null;
  columns: string[] | null;
  rows: Record<string, unknown>[] | null;
  row_count: number | null;
  plan: string[];
  optimization_notes: string[];
  explanation: string;
  source: "llm" | "fallback" | null;
  error: string | null;
}

export interface QueryHistoryItem {
  id: string;
  question: string;
  sql: string | null;
  explanation: string | null;
  source: string | null;
  row_count: number | null;
  error: string | null;
  created_at: string;
}

export async function getSchema(connectionId: string): Promise<SchemaTable[]> {
  const { data } = await api.get<{ tables: SchemaTable[] }>(
    `/connections/${connectionId}/schema`,
  );
  return data.tables;
}

export async function queryConnection(
  connectionId: string,
  question: string,
): Promise<NLQueryResponse> {
  const { data } = await api.post<NLQueryResponse>(
    `/connections/${connectionId}/query`,
    { question },
  );
  return data;
}

export async function listQueries(
  connectionId: string,
): Promise<QueryHistoryItem[]> {
  const { data } = await api.get<QueryHistoryItem[]>(
    `/connections/${connectionId}/queries`,
  );
  return data;
}

export async function deleteQuery(
  connectionId: string,
  queryId: string,
): Promise<void> {
  await api.delete(`/connections/${connectionId}/queries/${queryId}`);
}

File: backend/app/schemas/nl_sql.py
"""Pydantic schemas for NL->SQL over connected databases."""
import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class ColumnInfo(BaseModel):
    name: str
    type: str


class TableSchema(BaseModel):
    table: str
    columns: list[ColumnInfo]


class SchemaResponse(BaseModel):
    tables: list[TableSchema]


class NLQueryRequest(BaseModel):
    question: str = Field(min_length=1, max_length=2000)


class NLQueryResponse(BaseModel):
    sql: str | None = None
    columns: list[str] | None = None
    rows: list[dict[str, Any]] | None = None
    row_count: int | None = None
    plan: list[str] = []
    optimization_notes: list[str] = []
    explanation: str
    source: Literal["llm", "fallback"] | None = None
    error: str | None = None


class QueryHistoryItem(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    question: str
    sql: str | None = None
    explanation: str | None = None
    source: str | None = None
    row_count: int | None = None
    error: str | None = None
    created_at: datetime
